sydent/hs_federation/verifier.py

```diff
@@ -0,0 +1,184 @@
+            self.cache[server_name] = result
+
+        defer.returnValue(result['verify_keys'])
+
+    @defer.inlineCallbacks
+    def verifyServerSignedJson(self, signed_json, acceptable_server_names=None):
+        """Given a signed json object, try to verify any one
+        of the signatures on it
+        XXX: This contains a fairly noddy version of the home server
+        SRV lookup and signature verification. It only looks at
+        the first SRV result. It does no caching (just fetches the
+        signature each time and does not contact any other servers
+        to do perspectives checks.
+
+        :param acceptable_server_names: If provided and not None,
+        only signatures from servers in this list will be accepted.
+        :type acceptable_server_names: list of strings
+
+        :return a tuple of the server name and key name that was
+        successfully verified. If the json cannot be verified,
+        raises SignatureVerifyException.
+        """
+        if 'signatures' not in signed_json:
+            raise SignatureVerifyException("Signature missing")
+        for server_name, sigs in signed_json['signatures'].items():
+            if acceptable_server_names is not None:
+                if server_name not in acceptable_server_names:
+                    continue
+
+            server_keys = yield self._getKeysForServer(server_name)
+            for key_name, sig in sigs.items():
+                if key_name in server_keys:
+                    if 'key' not in server_keys[key_name]:
+                        logger.warn("Ignoring key %s with no 'key'")
+                        continue
+                    key_bytes = decode_base64(server_keys[key_name]['key'])
+                    verify_key = signedjson.key.decode_verify_key_bytes(key_name, key_bytes)
+                    logger.info("verifying sig from key %r", key_name)
+                    signedjson.sign.verify_signed_json(signed_json, server_name, verify_key)
+                    logger.info("Verified signature with key %s from %s", key_name, server_name)
+                    defer.returnValue((server_name, key_name))
+            logger.warn(
+                "No matching key found for signature block %r in server keys %r",
+                signed_json['signatures'], server_keys,
+            )
+        logger.warn(
+            "Unable to verify any signatures from block %r. Acceptable server names: %r",
+            signed_json['signatures'], acceptable_server_names,
+        )
+        raise SignatureVerifyException("No matching signature found")
+
+    @defer.inlineCallbacks
+    def authenticate_request(self, request, content):
+        """Authenticates a Matrix federation request based on the X-Matrix header
+        XXX: Copied largely from synapse
+
+        :param request: The request object to authenticate
+        :param content: The content of the request, if any
+        :type content: bytes or None
+        """
```

Comment:
could you document the return val please

Reply:
done